Created initial classes and tests

XedinUnknown opened the pull request:
These provide base concrete functionality for containers.

- Exception classes;
- Exception factories.

XedinUnknown committed:
Now depending on abstract and interface container packages

XedinUnknown committed:
Added classes and tests

XedinUnknown committed:
Interface package stability increased

All deps updated

XedinUnknown committed:
No longer depending on obsolete abstract package

XedinUnknown committed:
Re-factored and fixed

- Removed dependency on obsolete abstract classes.
- Added container-aware trait with test.
- Fixed and standardized tests.

XedinUnknown committed:
Code standards fixed

XedinUnknown committed:
Added `DataKeyAwareTrait` with test

XedinUnknown committed:
NetBeans project now targets PHP 5.4

XedinUnknown committed:
Added test for `ContainerAwareTrait` to make sure it fails correctly

XedinUnknown committed:
Travis will now build on Precise distro

XedinUnknown commented:
@mecha, could you please review this?

XedinUnknown committed:
Travis will no longer build with PHP 5.3

mecha commented on src/Exception/AbstractBaseContainerException.php:
```diff
@@ -0,0 +1,45 @@
+use Dhii\Data\Container\ContainerAwareTrait;
+
+/**
+ */
```
Missing class description and `since` tag.

mecha commented on src/ContainerAwareTrait.php:
```diff
@@ -0,0 +1,88 @@
+     * @param ContainerInterface|null $container The container.
+     *
+     * @throws InvalidArgumentException If not a valid container.
+     */
```
Missing `return` tag

XedinUnknown replied:
Here, absence of the `@return` tag indicates that the returned value is unknown, and should not be relied on.

mecha replied:
The return type is known: it is the instance the method is called on. FYI: Setters in traits _can_ have `@return $this`.

But I understand why you would not want to document this. In general, I think returning chaining is not something we should be doing within traits. Thoughts? If you agree, kindly remove the `return $this` at line 53.

XedinUnknown replied:
I agree. In fact, I think that chaining is not something we should be doing at all. I do, however, leave the opportunity for other implementations to implement chaining, which is why the type is undocumented, instead of documenting `null` or `void`.

mecha commented on src/DataKeyAwareTrait.php:
```diff
@@ -0,0 +1,88 @@
+     *
+     * @param string|Stringable|null $key The key.
+     *
+     * @return $this
```
Missing `throws` tag

XedinUnknown replied:
Actually, my idea is that perhaps it should become a standard that any method that accepts an argument can throw an `InvalidArgumentException`, and we can avoid documenting this.

XedinUnknown committed:
Fixed undeclared `InvalidArgumentException` usage

XedinUnknown committed:
Added class documentation

XedinUnknown committed:
Switched to task branch of interface package for testing

This is needed to test compliance with pending fix for dhii/data-container-interface#3

XedinUnknown commented:
Added fix for pending dhii/data-container-interface#3.

XedinUnknown committed:
Fixed parameters and docs

- `Stringable` is now accepted together with string where appropriate.
- Possibility of passing `null` documented.
- Added missing parameter docs to `AbstractContainerBase`.

XedinUnknown committed:
Fixed exception factories' docs

Summary had inheritdoc tag, probably doe to the fact that this class
used to extend an abstract.

mecha reviewed (changes requested):
@XedinUnknown Note that the `dhii/exception-helper-base` dependency is no longer available at version `dev-task/initial-classes`. Kindly updated to `dev-develop`

Also, a reminder of my [last comment](https://github.com/Dhii/data-container-base/pull/1#discussion_r136278962).

XedinUnknown committed:
Exception factories now allow `null` for inner exception

XedinUnknown committed:
Removed documentation that allows method chaining

XedinUnknown committed:
Switched to more recent versions of dependencies

This should fix broken builds, coupled with changes to the interface package.
